packages/menu/src/Menu.tsx

```diff
@@ -88,46 +89,158 @@ function Menu({
   ...rest
 }: MenuProps) {
   const refs: Array<HTMLElement> = [];
+  const titleArr: Array<string> = [];
+
   const hasSetInitialFocus = useRef(false);
+  const hasSetInitialOpen = useRef(false);
+  const [, setClosed] = useState(false);
+  const [currentSubMenu, setCurrentSubMenu] = useState<React.ReactElement<
+    HTMLLIElement
+  > | null>(null);
   const { setUsingKeyboard } = useUsingKeyboardContext();
 
-  const updatedChildren = React.Children.map(children, child => {
-    if (
-      (isComponentType(child, 'MenuItem') && !child.props.disabled) ||
-      isComponentType(child, 'FocusableMenuItem')
-    ) {
-      return React.cloneElement(child, {
-        ref: (ref: HTMLElement) => {
-          refs.push(ref);
-          if (
-            open &&
-            refs.length === 1 &&
-            hasSetInitialFocus.current === false
-          ) {
-            refs[0].focus();
-            setUsingKeyboard(true);
-            setFocused(refs[0]);
-            hasSetInitialFocus.current = true;
-          }
-        },
-        onFocus: (e: Event) => setFocused(e.target as HTMLElement),
-      });
-    }
+  function updateChildren(
+    children: Array<React.ReactElement>,
+  ): Array<React.ReactElement> {
+    return React.Children.map(children, (child: React.ReactElement) => {
+      if (child.props?.disabled) {
+        return child;
+      }
 
-    return child;
-  });
+      let currentChildRef: HTMLElement | null = null;
+
+      const setRef = (ref: HTMLElement) => {
+        if (ref == null) {
+          return;
+        }
+
+        refs.push(ref);
+        currentChildRef = ref;
+
+        if (open && hasSetInitialFocus.current === false) {
+          setFocus(refs[0]);
+          // This allows the focus state to be shown via the LeafyGreen Provider
+          setUsingKeyboard(true);
+          hasSetInitialFocus.current = true;
+        }
+      };
+
+      const title = child?.props?.title ?? false;
+
+      const onFocus = ({ target }: { target: HTMLElement }) => {
+        setFocused(target);
+      };
+
+      if (isComponentType<'SubMenu'>(child, 'SubMenu') && title) {
+        if (titleArr.includes(title)) {
+          throw new Error('SubMenu titles must be unique');
+        }
+
+        titleArr.push(title);
+
+        if (child.props.active && !hasSetInitialOpen.current) {
+          setCurrentSubMenu(child);
+          hasSetInitialOpen.current = true;
+        }
+
+        const isCurrentSubMenu = currentSubMenu?.props.title === title;
+
+        return React.cloneElement(child, {
+          ref: setRef,
+          open: isCurrentSubMenu,
```

Comment:
Not a blocker, but we might consider reordering these for readability, putting all the inline function declarations last.